android/java/org/chromium/chrome/browser/vpn/activities/BraveVpnParentActivity.java

```diff
@@ -0,0 +1,194 @@
+import org.chromium.chrome.browser.vpn.utils.BraveVpnUtils;
+import org.chromium.chrome.browser.vpn.utils.InAppPurchaseWrapper;
+import org.chromium.chrome.browser.vpn.wireguard.WireguardConfigUtils;
+import org.chromium.chrome.browser.vpn.wireguard.WireguardService;
+import org.chromium.ui.widget.Toast;
+
+import java.util.List;
+import java.util.TimeZone;
+
+public abstract class BraveVpnParentActivity
+        extends AsyncInitializationActivity implements BraveVpnObserver {
+    public boolean mIsVerification;
+    private BraveVpnPrefModel mBraveVpnPrefModel;
+
+    abstract void showRestoreMenu(boolean shouldShowRestore);
+    abstract void updateProfileView();
+
+    ActivityResultLauncher<Intent> intentActivityResultLauncher = registerForActivityResult(
+            new ActivityResultContracts.StartActivityForResult(), result -> {
+                BraveVpnUtils.dismissProgressDialog();
+                if (result.getResultCode() == RESULT_OK) {
+                    BraveVpnProfileUtils.getInstance().startVpn(BraveVpnParentActivity.this);
+                    BraveVpnUtils.showVpnConfirmDialog(this);
+                } else if (result.getResultCode() == RESULT_CANCELED) {
+                    if (BraveVpnProfileUtils.getInstance().isVPNRunning(this)) {
+                        BraveVpnUtils.showVpnAlwaysOnErrorDialog(this);
+                    } else {
+                        updateProfileView();
+                    }
+                    Toast.makeText(this, "Permission cancelled", Toast.LENGTH_SHORT).show();
+                }
+            });
+
+    @Override
+    public void finishNativeInitialization() {
+        super.finishNativeInitialization();
+        if (BraveVpnUtils.isBraveVpnFeatureEnable()) {
+            InAppPurchaseWrapper.getInstance().startBillingServiceConnection(
+                    BraveVpnParentActivity.this);
+        }
+    }
+
+    protected void verifySubscription() {
+        mBraveVpnPrefModel = new BraveVpnPrefModel();
+        List<Purchase> purchases = InAppPurchaseWrapper.getInstance().queryPurchases();
+        if (purchases != null && purchases.size() == 1) {
+            Purchase purchase = purchases.get(0);
+            mBraveVpnPrefModel.setPurchaseToken(purchase.getPurchaseToken());
+            mBraveVpnPrefModel.setProductId(purchase.getSkus().get(0).toString());
+            BraveVpnNativeWorker.getInstance().verifyPurchaseToken(
+                    mBraveVpnPrefModel.getPurchaseToken(), mBraveVpnPrefModel.getProductId(),
+                    BraveVpnUtils.SUBSCRIPTION_PARAM_TEXT, getPackageName());
+        } else {
+            if (!mIsVerification) {
+                BraveVpnApiResponseUtils.queryPurchaseFailed(BraveVpnParentActivity.this);
+            } else {
+                BraveVpnUtils.dismissProgressDialog();
+            }
+        }
+    }
+
+    @Override
+    public void onVerifyPurchaseToken(String jsonResponse, boolean isSuccess) {
```

Comment:
Should this method be reentrant?

Reply:
Same as above. this callback has to be sequential as we rely on the response and start another api call if we need to.